### PR TITLE
[RLlib] APPO enhancements (new API stack) vol 01: Add circular buffer

### PR DESCRIPTION
**Pull request by sven1977**
APPO enhancements (new API stack) vol 01: Add circular buffer

Adds the circular buffer described in the [IMPACT paper](https://arxiv.org/pdf/1912.00167) to the (new stack) implementation.

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: rllib/algorithms/appo/appo.py [hunk 1]
```diff
@@ -193,6 +198,14 @@ def training(
                 on before updating the target networks and tune the kl loss
                 coefficients. NOTE: This parameter is only applicable when using the
                 Learner API (enable_rl_module_and_learner=True).
+            circular_buffer_num_batches: The number of train batches that fit
+                into the circular buffer. Each such train batch can be sampled for
+                training max. `circular_buffer_iterations_per_batch_K` times.
```
Comment (simonsays1980) [inline rllib/algorithms/appo/appo.py:203]: The `circular_buffer_iterations_per_batch_K` is irritating as it is desribed nowhere. Instead we have a very similar attribute below.

File: rllib/algorithms/appo/appo.py [hunk 2]
```diff
@@ -227,9 +240,40 @@ def training(
...
+                    "any batch may be sampled, set "
+                    "`circular_buffer_iterations_per_batch`."
+                )
+            if self.num_multi_gpu_tower_stacks != 1:
+                raise ValueError(
+                    "`num_multi_gpu_tower_stacks` not supported on new API stack with "
+                    "APPO! In order to train on multi-GPU, use "
+                    "`config.learners(num_learners=[number of GPUs], "
+                    "num_gpus_per_learner=1)`. To scale the throughput of batch-to-GPU-"
+                    "pre-loading on each of your `Learners`, set "
```
Comment (simonsays1980) [inline rllib/algorithms/appo/appo.py:272]: Super nice descriptions! This helps users! We might even make an extra section in the docs for "Tuning your applications" where we describe for different algorithms some scenarios.

File: rllib/algorithms/appo/appo_learner.py [hunk 3]
```diff
@@ -28,6 +29,11 @@ class APPOLearner(IMPALALearner):
 
     @override(IMPALALearner)
     def build(self):
+        self._learner_thread_in_queue = CircularBuffer(
+            num_batches=self.config.circular_buffer_num_batches_N,
+            iterations_per_batch=self.config.circular_buffer_iterations_per_batch_K,
```
Comment (simonsays1980) [inline rllib/algorithms/appo/appo_learner.py:34]: Ah okay, so it does exist :)

File: rllib/algorithms/appo/utils.py [hunk 4]
```diff
@@ -1,12 +1,89 @@
...
 
 
-# TODO (sven): Deprecate once APPO and IMPALA fully on RLModules/Learner APIs.
+class CircularBuffer:
+    """A circular batch-wise buffer as described in [1] for APPO.
+
+    The buffer holds at most N batches, which are sampled at random (uniformly).
+    If full and a new batch is added, the oldest batch is discarded. Also, each batch
+    currently in the buffer can be sampled at most K times (after which it is also
+    discrded).
```
Comment (simonsays1980) [inline rllib/algorithms/appo/utils.py:26]: discrded -> discarded ;)

File: rllib/algorithms/appo/utils.py [hunk 5]
```diff
@@ -1,12 +1,89 @@
...
+        # Add buffer and k=0 information to the deque.
+        with self._lock:
+            len_ = len(self._buffer)
+            if len_ == self.num_batches:
+                dropped_entry = self._buffer[0]
+            self._buffer.append([batch, 0])
+
+        # A valid entry (w/ a batch whose k has not been reach K yet) was dropped.
+        if dropped_entry is not None and dropped_entry[0] is not None:
+            dropped_ts += dropped_entry[0].env_steps() * (
```
Comment (simonsays1980) [inline rllib/algorithms/appo/utils.py:51]: Does this still rely on `MultiAgentBatch`es instead episodes? Or has it already passed the learner connector?

File: rllib/algorithms/appo/utils.py [hunk 6]
```diff
@@ -1,12 +1,89 @@
...
+
+        return dropped_ts
+
+    def sample(self):
+        k = entry = batch = None
+
+        while True:
+            # Only initially, the buffer may be empty -> Just wait for some time.
+            if len(self._buffer) == 0:
+                time.sleep(0.001)
```
Comment (simonsays1980) [inline rllib/algorithms/appo/utils.py:63]: So rare to find something like this in code :D

File: rllib/algorithms/appo/utils.py [hunk 7]
```diff
@@ -1,12 +1,89 @@
...
+                continue
+            # Sample a random buffer index.
+            with self._lock:
```
Comment (simonsays1980) [inline rllib/algorithms/appo/utils.py:66]: We use the locks such that the buffer is not accessed by multiple threads while some are pushing in and others are sampling from, correct?

File: rllib/algorithms/appo/utils.py [hunk 8]
```diff
@@ -1,12 +1,89 @@
...
+                entry = self._buffer[random.randint(0, len(self._buffer) - 1)]
+            batch, k = entry
+            # Ignore batches that have already been invalidated.
+            if batch is not None:
+                break
+
+        # Increase k += 1 for this batch.
+        assert k is not None
+        entry[1] += 1
```
Comment (simonsays1980) [inline rllib/algorithms/appo/utils.py:75]: Sweet. Really simple and smooth implementation!

File: rllib/algorithms/impala/impala_learner.py [hunk 9]
```diff
@@ -175,15 +168,25 @@ def update_from_episodes(
...
-                        {mid: SampleBatch(b) for mid, b in batch.items()},
-                        env_steps=env_steps,
-                    ),
-                    self.metrics,
+                ma_batch = MultiAgentBatch(
+                    {mid: SampleBatch(b) for mid, b in batch.items()},
+                    env_steps=env_steps,
                 )
+                # Add the batch directly to the circular buffer.
+                if isinstance(self._learner_thread_in_queue, CircularBuffer):
```
Comment (simonsays1980) [inline rllib/algorithms/impala/impala_learner.py:176]: If the batch has already passed the learner connector, could there be any transformations in the connector that were made with a module that could become stale after some time of the batch in the buffer?

File: rllib/algorithms/impala/impala_learner.py [hunk 10]
```diff
@@ -263,8 +266,17 @@ def _step(self) -> None:
...
             )
-            # Enqueue to Learner thread's in-queue.
-            _LearnerThread.enqueue(self._out_queue, ma_batch_on_gpu, self.metrics)
+
+            if isinstance(self._out_queue, CircularBuffer):
+                ts_dropped = self._out_queue.add(ma_batch_on_gpu)
+                self.metrics.log_value(
+                    (ALL_MODULES, LEARNER_THREAD_ENV_STEPS_DROPPED),
+                    ts_dropped,
+                    reduce="sum",
```
Comment (simonsays1980) [inline rllib/algorithms/impala/impala_learner.py:275]: So nice with the different aggregation methods. Just put it in and aggregate.

**Review by simonsays1980 — APPROVED** [review]
LGTM. Awesome changes. Some small nit.

File: rllib/algorithms/appo/utils.py (hunk 6 above)
Comment (sven1977) [reply rllib/algorithms/appo/utils.py:63]: :) This sleep will only ever be used at the very beginning, when the buffer is still empty, but the Learner thread already starts sampling from it.

File: rllib/algorithms/appo/utils.py (hunk 4 above)
Comment (sven1977) [reply rllib/algorithms/appo/utils.py:26]: fixed

File: rllib/algorithms/impala/impala_learner.py (hunk 10 above)
Comment (sven1977) [reply rllib/algorithms/impala/impala_learner.py:275]: Yeah, that's the idea: Push and forget.

File: rllib/algorithms/impala/impala_learner.py (hunk 9 above)
Comment (sven1977) [reply rllib/algorithms/impala/impala_learner.py:176]: Could be. I think that's by design, though (and v-trace is there for the defense :) ).
For example the paper talks about doing the target net forward pass on the batch already before adding it to the buffer. However, we don't do this, we perform a new target forward pass each time we compute a loss. I don't think it's crucial, though, that we do it this way. We would save a few (cheap, non-gradient) forward passes.

**Pull request merged**